### ISSUE TITLE
Idea: Generic key type

### ISSUE THREAD

**Issue by jabolopes** [marwan-at-work/singleflight#1]
Hi,

Great package! An interesting feature would be to make the key type also generic, instead of always being a string. For example, for callers wanting to use an int as key they wouldn't have to convert  the int to a string.

Best regards,
Jose

**Comment by marwan-at-work** [comment]
That's a good idea. Though it makes me wonder why the original package not do the same by making the key also an `interface{}` which is comparable? 

Here's what the code change looks like: https://github.com/marwan-at-work/singleflight/compare/generickey?expand=1

It does make the constructor a little more verbose: `var g singleflight.Group[string, string]` but happy to merge that in if it's useful.

**Comment by jabolopes** [comment]
> It does make the constructor a little more verbose: var g singleflight.Group[string, string]

Yes, that's true. I guess it would be useful to be able to specify defaults for the template variables, e.g.,
```go
type Group[T = string comparable] ...
```

It would also mean this would be a backwards compatible change. But AFAIK this is not yet possible.
